libs/pdi-scanner/src/protocol/parsers.rs

```diff
@@ -0,0 +1,1297 @@
+/// ```
+///
+/// # Errors
+///
+/// Returns an error if the input does not match the expected format.
+pub fn get_test_string_response(input: &[u8]) -> IResult<&[u8], &str> {
+    map_res(packet((tag(b"D"), packet_body)), |(_, test_string)| {
+        from_utf8(test_string)
+    })(input)
+}
+
+simple_request!(get_firmware_version_request, b"V");
+
+/// Parses the response to a firmware version request.
+///
+/// # Errors
+///
+/// Returns an error if the input does not match the expected format.
+pub fn get_firmware_version_response(input: &[u8]) -> IResult<&[u8], Version> {
+    map(
+        packet((
+            tag(b"V"),
+            map_res(take(4usize), |bytes| from_utf8(bytes)),
+            map_res(take(2usize), |bytes| from_utf8(bytes)),
+            map_res(take(2usize), |bytes| from_utf8(bytes)),
+            map_res(take(1usize), |bytes| from_utf8(bytes)),
+        )),
+        |(_, product_id, major, minor, cpld_version)| {
+            Version::new(
+                product_id.to_owned(),
+                major.to_owned(),
+                minor.to_owned(),
+                cpld_version.to_owned(),
+            )
+        },
+    )(input)
+}
+
+simple_request!(get_current_firmware_build_version_string_request, b"\x1bV");
+
+/// Parses the response to a current firmware build version string request.
+///
+/// # Errors
+///
+/// Returns an error if the input does not match the expected format.
+pub fn get_current_firmware_build_version_string_response(input: &[u8]) -> IResult<&[u8], &str> {
+    if let Ok(([], _)) = packet((
+        tag(b"X"),
+        take(3usize),
+        tag(b" "),
+        take(2usize),
+        tag(b" "),
+        take(4usize),
+        tag(b"/"),
+        take(2usize),
+        tag(b":"),
+        take(2usize),
+        tag(b":"),
+        take(2usize),
+    ))(input)
+    {
+        if let Some(body) = extract_packet_body(input) {
+            return from_utf8(&input[b"X".len()..])
+                .map(|string| (&[] as &[u8], string))
+                .map_err(|_| {
+                    nom::Err::Failure(nom::error::Error::new(body, nom::error::ErrorKind::Verify))
+                });
+        }
+    }
+
+    Err(nom::Err::Error(nom::error::Error::new(
+        input,
+        nom::error::ErrorKind::Eof,
+    )))
+}
+
+simple_request!(get_scanner_status_request, b"Q");
+
+/// Parses the response to a scanner status request.
+///
+/// # Errors
+///
+/// Returns an error if the input does not match the expected format.
+pub fn get_scanner_status_response(input: &[u8]) -> IResult<&[u8], Status> {
+    map(
+        packet((tag(b"Q"), le_u8, le_u8, le_u8)),
+        |(_, byte0, byte1, byte2)| {
+            Status::new(
+                byte0 & 0b0000_0001 != 0,
+                byte0 & 0b0000_0010 != 0,
+                byte0 & 0b0000_0100 != 0,
+                byte0 & 0b0000_1000 != 0,
+                byte0 & 0b0001_0000 != 0,
+                byte0 & 0b0100_0000 != 0,
```

Comment:
I noticed these two lines skip over `0b0010_0000`. Is that intentional?

Reply:
The "Combined PS5 UltraScan Software Spec.pdf" file defines the response to the Request Scanner Status ('Q') command in section 2.1.51. It lists the first byte, bit 5 as being reserved for future use.

Reply:
Ah, I see. Thanks for the pointer. Maybe a comment here would be nice to help remember